### ISSUE TITLE
AggregateRight

### ISSUE THREAD

**Issue by leandromoh** [morelinq/MoreLINQ#216]
It's similar to `Aggregate` but Enumerates from right to left, 

I'm planning to submit a PR for this. 

What do you think?

**Comment by fsateler** [comment]
This operator would be equivalent to `src.Reverse().Aggregate()`, wouldn't it? If not, what would be different?

**Comment by atifaziz** [comment]
Wondering the same here as @fsateler.

**Comment by leandromoh** [comment]
> This operator would be equivalent to src.Reverse().Aggregate(), wouldn't it? If not, what would be different?

If you program in Haskell, Lisp, Scala, Standard ML, Smalltalk, among others, it isn't equivalent.

If you use `Enumerable.Range(1, 13).Reverse().Aggregate((a, b) => string.Format("({0}/{1})", a, b))`  the result is:
((((((((((((13/12)/11)/10)/9)/8)/7)/6)/5)/4)/3)/2)/1)

Whereas if you also program in one of above languages you will expect:
(1/(2/(3/(4/(5/(6/(7/(8/(9/(10/(11/(12/13))))))))))))

Although the parenteses order is right, the order of the arguments is not, what will give you a complete unexpected return value; 

If you are passing a lambda as argument of course you can change the arguments order within it (e.g.,`(a, b) => string.Format("({0}/{1})", b, a)` in the original example). but for functions already declared (e.g., class members, framework's functions, etc) you will need to create a wrapper (e.g., `(b, a) => f(a, b)`). It works, but of course `source.Reverse().Aggregate((a, b) => f(b,a))` is much more ugly than simply call `source.AggregateRight(f)`.

In short, make sense to me create the `AggregateRight` function the way I mentioned. Other point, but lesser important, is that if the collection implements `IList<T>`, is much more performatic to make a reverse `for `than call `Reverse()`.

**Comment by atifaziz** [comment]
> It's similar to `Aggregate` but Enumerates from right to left,

@leandromoh So glad to hear you mean like `foldr` (below) where it's about _right-associativity. Just wanted to be sure it's not a simple case of enumerating from right to left as you said in your original description.

![foldr](https://upload.wikimedia.org/wikipedia/commons/3/3e/Right-fold-transformation.png)

It's a welcomed addition, especially with respect to optimization over lists, so feel free to submit your PR.

**Comment by fsateler** [comment]
> Although the parenteses order is right, the order of the arguments is
not, what will give you a complete unexpected return value;

Ah, thanks for explaining!

**Comment by leandromoh** [comment]
PR submitted

**Comment by atifaziz** [comment]
Let's try to get this into [2.2](https://github.com/morelinq/MoreLINQ/milestone/3) at end of month.
